Compiling Errors with C++20 in NN_NO_CHECK

I compile the package with msvc 2019 using the C++20 (i.e., ISO C++20 standard (/std:c++20)), and some errors occured.

It indicated problems from "NN_NO_CHECK"

#define NN_NO_CHECK(p)                                                         \
    ::dropbox::oxygen::nn<typename std::remove_const<                          \
        typename std::remove_reference<decltype(p)>::type>::type>(             \
        dropbox::oxygen::i_promise_i_checked_for_null, (p))

How should I resolve these problems?

Thanks!

1>D:\opensource\gdalvs2019\proj-9.2.1(c20)\src\iso19111\operation\coordinateoperationfactory.cpp(3777,49): error C2039: "type": is not a member of  "std" 
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include\algorithm(40): message : 参见“std”的声明
1>D:\opensource\gdalvs2019\proj-9.2.1(c20)\src\iso19111\operation\coordinateoperationfactory.cpp(3800,36): error C2039: "type": is not a member of  "std" 
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include\algorithm(40): message : 参见“std”的声明
1>D:\opensource\gdalvs2019\proj-9.2.1(c20)\src\iso19111\operation\coordinateoperationfactory.cpp(3844,23): error C2039: "type": is not a member of  "std" 
……




<!-- Please search existing issues to avoid creating duplicates. -->


#### Example of problem
Demonstrate the problem you have found, either by using PROJ tools like `cs2cs`, `cct` or `projinfo` **or** add a code snippet that highlights the problem using the PROJ C/C++ API.
A "Minimal, Complete and Verifiable Example" will make it much easier for maintainers to help you:
http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

```c
# Your code here

```
#### Problem description

[this should explain **why** the current behavior is a problem and why the expected output is a better solution.]

#### Expected Output


#### Environment Information
 - PROJ version (`proj`)
 - Operation System Information

#### Installation method
 - conda, apt-get, from source, etc...

--- a/.github/workflows/windows.yml
+++ b/.github/workflows/windows.yml
@@ -61,7 +61,9 @@ jobs:
             mkdir %PROJ_BUILD%
             cd %PROJ_BUILD%
             set PROJ_DIR=%GITHUB_WORKSPACE%\proj_dir
-            cmake -DCMAKE_BUILD_TYPE="${{ env.BUILD_TYPE }}" -DBUILD_SHARED_LIBS="${{ env.BUILD_SHARED_LIBS }}" -DCMAKE_C_FLAGS="/WX" -DCMAKE_CXX_FLAGS="/WX" -DCMAKE_TOOLCHAIN_FILE=c:/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_INSTALL_PREFIX="%PROJ_DIR%" -DPROJ_DB_CACHE_DIR=%PROJ_DB_CACHE_DIR% ..
+            # Not directly linked to BUILD_SHARED_LIBS, but a way to test different C++ standard versions
+            if "${{ env.BUILD_SHARED_LIBS }}"=="ON" (set EXTRA_CXX_FLAGS="/std:c++20")
+            cmake -DCMAKE_BUILD_TYPE="${{ env.BUILD_TYPE }}" -DBUILD_SHARED_LIBS="${{ env.BUILD_SHARED_LIBS }}" -DCMAKE_C_FLAGS="/WX" -DCMAKE_CXX_FLAGS="/WX %EXTRA_CXX_FLAGS%" -DCMAKE_TOOLCHAIN_FILE=c:/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_INSTALL_PREFIX="%PROJ_DIR%" -DPROJ_DB_CACHE_DIR=%PROJ_DB_CACHE_DIR% ..
             ninja -v
             ninja install
             dir %PROJ_DIR%\bin

--- a/include/proj/util.hpp
+++ b/include/proj/util.hpp
@@ -42,6 +42,7 @@
 #include <map>
 #include <memory>
 #include <string>
+#include <type_traits>
 #include <vector>
 
 #ifndef NS_PROJ
@@ -203,10 +204,25 @@ using ::dropbox::oxygen::nn_static_pointer_cast;
 
 template <typename T> using nn_shared_ptr = nn<std::shared_ptr<T>>;
 
+// Possible implementation of C++14 std::remove_reference_t
+// (cf https://en.cppreference.com/w/cpp/types/remove_cv)
+template <class T>
+using remove_reference_t = typename std::remove_reference<T>::type;
+
+// Possible implementation of C++14 std::remove_cv_t
+// (cf https://en.cppreference.com/w/cpp/types/remove_cv)
+template <class T> using remove_cv_t = typename std::remove_cv<T>::type;
+
+// Possible implementation of C++20 std::remove_cvref
+// (cf https://en.cppreference.com/w/cpp/types/remove_cvref)
+template <class T> struct remove_cvref {
+    typedef remove_cv_t<remove_reference_t<T>> type;
+};
+
 #define NN_NO_CHECK(p)                                                         \
-    ::dropbox::oxygen::nn<typename std::remove_const<                          \
-        typename std::remove_reference<decltype(p)>::type>::type>(             \
-        dropbox::oxygen::i_promise_i_checked_for_null, (p))
+    ::dropbox::oxygen::nn<                                                     \
+        typename ::NS_PROJ::util::remove_cvref<decltype(p)>::type>(            \
+        ::dropbox::oxygen::i_promise_i_checked_for_null, (p))
 
 //! @endcond
 

